File: src/dataset/javascript/forth_exercism-ThomasZumsteg/gemini/GeminiFlash_forth_v3.js
function n_args(n, func) {
  return function(stack) {
    if (stack.length < n) {
      throw new Error("Stack empty");
    }
    const args = stack.splice(stack.length - n);
    stack.push(...func(...args));
  };
}

export default class Forth {
  constructor() {
    this.stack = [];
    this.env = {};
    this.builtin = {
      "+": n_args(2, (a, b) => [a + b]),
      "-": n_args(2, (a, b) => [a - b]),
      "*": n_args(2, (a, b) => [a * b]),
      "/": n_args(2, (a, b) => {
        if (b === 0) {
          throw new Error("Division by zero");
        }
        return [Math.trunc(a / b)];
      }),
      "dup": n_args(1, a => [a, a]),
      "drop": n_args(1, a => []),
      "swap": n_args(2, (a, b) => [b, a]),
      "over": n_args(2, (a, b) => [a, b, a]),
    };
  }

  evaluate(line) {
    if (line.startsWith(': ') && line.endsWith(' ;')) {
      const tokens = line.slice(2, -2).split(" ");
      const head = tokens[0].toLowerCase();
      if (!isNaN(Number(head))) {
        throw new Error("Invalid definition");
      }
      this.env[head] = tokens.slice(1);
      return;
    }

    const tokens = line.split(" ");
    for (let i = 0; i < tokens.length; i++) {
      const token = tokens[i].toLowerCase();
      const num = Number(token);
      if (!isNaN(num)) {
        this.stack.push(num);
      } else if (this.env.hasOwnProperty(token)) {
        tokens.splice(i + 1, 0, ...this.env[token]);
      } else if (this.builtin.hasOwnProperty(token)) {
        this.builtin[token](this.stack);
      } else {
        throw new Error('Unknown command');
      }
    }
  }
}

File: src/dataset/python/forth_exercism-ThomasZumsteg/gemini/GeminiFlash_forth_v3.py
class StackUnderflowError(Exception):
    pass


keywords = {
    '+': (2, lambda a, b: (a + b,)),
    '-': (2, lambda a, b: (a - b,)),
    '*': (2, lambda a, b: (a * b,)),
    '/': (2, lambda a, b: (a // b,)),
    'swap': (2, lambda a, b: (b, a)),
    'over': (2, lambda a, b: (a, b, a)),
    'dup': (1, lambda x: (x, x)),
    'drop': (1, lambda x: tuple()),
}


def evaluate(input_data):
    stack = []
    env = {}
    for line in input_data:
        tokens = line.split()
        i = 0
        while i < len(tokens):
            token = tokens[i].lower()
            if token == ':':
                i += 1
                keyword = tokens[i].lower()
                if keyword.isdigit():
                    raise ValueError("Variables cannot be numbers: " + keyword)
                i += 1
                env[keyword] = []
                while tokens[i] != ';':
                    env[keyword].append(tokens[i])
                    i += 1
                i += 1
            elif token in env:
                tokens = env[token] + tokens[i+1:]
                i = 0
            elif token in keywords:
                n_args, func = keywords[token]
                if len(stack) < n_args:
                    raise StackUnderflowError('Cannot apply ' + token)
                args = tuple(stack[-n_args:][::-1])
                del stack[-n_args:]
                stack.extend(func(*args))
                i += 1
            else:
                try:
                    stack.append(int(token))
                except ValueError:
                    raise ValueError(f"Invalid token: {token}")
                i += 1
    return stack